File: Leetcode/array_hashing/python3/group-anagrams.py
from typing import List
from collections import Counter


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        if len(strs) == 1:
            return [strs]
        else:
            anagramDict = {}
            for word in strs:
                sortedWord = "".join(sorted(word))
                if sortedWord in anagramDict:
                    anagramDict[sortedWord].append(word)
                else:
                    anagramDict[sortedWord] = [word]
            return list(anagramDict.values())

File: Leetcode/array_hashing/scala/group-anagrams.scala
object Solution {
    import scala.collection.mutable._

    def groupAnagrams(strs: Array[String]): List[List[String]] = {
        val map = new HashMap[String, ListBuffer[String]]()

        for (word <- strs) {
            val sortedWord = word.sorted

            if (!map.contains(sortedWord)) {
                map(sortedWord) = new ListBuffer[String]()
            }

            map(sortedWord).append(word)
        }

        map.values.map(_.toList).toList 
    }
}